File: csharp/matching-brackets/MatchingBrackets.cs
using System;
using System.Collections.Generic;

public static class MatchingBrackets
{
    private static IDictionary<char, char> ClosingChar = new Dictionary<char, char>
    {
        { ']', '['},
        { '}', '{'},
        { ')', ')'}
    };


    public static bool IsPaired(string input)
    {
        var stack = new Stack<char>();

        foreach(var c in input)
        {

            switch(c)
            {
                case '[':
                case '{':
                case '(':
                    stack.Push(c);
                    break;
                case ']':
                case '}':
                case ')':
                    if (stack.Count > 0 && stack.Peek() == ClosingChar[c])
                    {
                        stack.Pop();
                    }
                    else
                    {
                        return false;
                    }
                    break;
            }
        }

        return stack.Count == 0;
    }
}

File: javascript/matching-brackets/matching-brackets.js
const closingChar = new Map([
  ["]", "["],
  ["}", "{"],
  [")", "("],
]);

// Will have already checked length > 0 before calling
const peek = (stack) => stack[stack.length - 1];

const isStackEmpty = stack => stack.length == 0;

export const isPaired = (text) => {

  let stack = [];

  for (const c of text) {

    switch (c) {
      case "[":
      case "{":
      case "(":
        stack.push(c);
        break;
      case "]":
      case "}":
      case ")":
        if (stack.length > 0 && peek(stack) == closingChar.get(c)) {
          stack.pop();
        } else {
          return false;
        }
        break;
    }
  }

  return isStackEmpty(stack);
};
